[Feature] Update Node.js version from v12 to v20 as GitHub Actions will no longer support v12

danielmarcano opened the pull request:
Related to Issue [#9895](https://github.com/Submitty/Submitty/issues/9895)

The following changes have been made because GitHub Actions will no longer support Node.js v12, as mentioned by @williamjallen in the linked issue. You can find the relevant information in the blog post referenced in that issue.

1. Updated `publish.yml` so that it uses Node.js version `20`
2. Updated `test.yml` so that it uses Node.js version `[20.x]`
3. Updated `version` to `2.5.2` in `package.json` to reflect this change

### Checks:

- [x] Have run `npm run lint` and `npm run test` as specified in the `README.md`

P.S. If this PR is accepted, could you please add the "hacktoberfest-accepted" label to it, so that it counts towards my Hacktoberfest contributions?

williamjallen commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@submitty/pdf-annotate.js",
-  "version": "2.5.1",
+  "version": "2.5.2",
```
Is this change and the associated changes in `package-lock.json` intentional?

williamjallen replied:
These changes should be reverted, and made as a separate version bump PR when a new release is made.

danielmarcano replied:
Hi, @williamjallen. They are intentional, as the Node.js version is not trivial.

I have actually just noticed that the GitHub Action is failing due to the Node.js version increase. I will take a look at it tomorrow.

If you still consider that a version increase should be done in a different PR, do let me know, and I will revert the change.

williamjallen replied:
See [this](https://github.com/Submitty/pdf-annotate.js/commit/6edea3993532e4bf3a374a8d249e15edf13f2f39) commit for an example of how our release process typically works in this repository.

The test needs to be fixed, and the version changes need to be reverted, and then this PR should be ready to be merged.

danielmarcano replied:
Thank you! I will check it out and revert the version change.

danielmarcano commented:
Hi, @williamjallen, I have figured out that the issue was using Webpack v4, as it does not support Node.js v20.

Thus, I have updated all Webpack, Webpack Plugins, and Karma packages to their latest versions so that they can support Node.js v20.

I have also updated Webpack and Karma configs to meet new Webpack v5 requirements and syntax.

Nonetheless, there's a small issue. Four tests are failing. I have tried my best to fix them but could not understand how to do so.

I have also ran the project with `npm start` and the failing tests features are indeed working:

- Add annotation
- Edit annotation
- Delete annotation

<img width="378" alt="image" src="https://github.com/Submitty/pdf-annotate.js/assets/93147810/c184da29-14f8-4e6c-bda8-104fa69ffc1c">

Could you guide me on how to fix these tests?

Thank you.

williamjallen commented:
Does adding a `disableEdit()` at the end help?  It looks a little bit odd to have timeouts all over the place as well, but cleaning up all of that isn't necessarily something we should fix in this PR.

I'm not particularly knowledgeable about this codebase (or Webpack in general for that matter).  If you post in one of the Submitty Slack channels, someone who has worked with this repository might be able to provide more assistance.

danielmarcano commented:
> disableEdit

I tried your suggestion, but it did not work. For now, I have skipped that sole test, and left a comment explaining why.

If you think we would be good to go, please take a look when you can and approve it if you think everything's okay.
